### ISSUE TITLE
No connection pool for 'ActiveRecord::Base' found

### ISSUE THREAD

**Issue by morgoth** [bensheldon/good_job#236]
I start good job, by `bundle exec good_job start --poll-interval=1`

for first few jobs that are enqueued, I got the stacktrace:
```
[GoodJob] GoodJob error: No connection pool for 'ActiveRecord::Base' found.
"/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/connection_handling.rb:327:in `retrieve_connection'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/connection_handling.rb:283:in `connection'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/delegation.rb:93:in `connection'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/query_methods.rb:1394:in `preprocess_order_args'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/query_methods.rb:360:in `order!'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/query_methods.rb:355:in `order'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/good_job-1.8.0/lib/good_job/job.rb:75:in `block in <class:Job>'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:411:in `instance_exec'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:411:in `block in _exec_scope'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:810:in `_scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:411:in `_exec_scope'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/scoping/named.rb:176:in `block in scope'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/delegation.rb:67:in `block in priority_ordered'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:406:in `block in scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:810:in `_scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:406:in `scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/delegation.rb:67:in `priority_ordered'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/good_job-1.8.0/lib/good_job/job.rb:147:in `perform_with_advisory_lock'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/delegation.rb:67:in `block in perform_with_advisory_lock'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:406:in `block in scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:810:in `_scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation.rb:406:in `scoping'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.1/lib/active_record/relation/delegation.rb:67:in `perform_with_advisory_lock'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/good_job-1.8.0/lib/good_job/job_performer.rb:29:in `next'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/good_job-1.8.0/lib/good_job/scheduler.rb:215:in `block (2 levels) in create_task'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activesupport-6.1.3.1/lib/active_support/execution_wrapper.rb:88:in `wrap'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/good_job-1.8.0/lib/good_job/scheduler.rb:215:in `block in create_task'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:24:in `block in execute'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `block in synchronize'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:19:in `execute'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/ivar.rb:169:in `safe_execute'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/scheduled_task.rb:285:in `process_task'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/timer_set.rb:98:in `block in ns_post_task'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:363:in `run_task'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:352:in `block (3 levels) in create_worker'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `loop'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `block (2 levels) in create_worker'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `catch'",
 "/home/wojtek/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `block in create_worker'"
```

The jobs are processed fine and there is no other error thrown.

I have multiple database configured in my Rails 6.1.3.1 app, so maybe it is caused by this.
```
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true

  connects_to database: {writing: :primary}
end
```

Also at the beginning I can see in logs;
```
[GoodJob] [2888577] [GoodJob::Scheduler(queues=* max_threads=5)-worker-2] Executed GoodJob cc907d6d-30d8-483e-ad12-be17d2dd0011
WARNING:  you don't own a lock of type ExclusiveLock
[GoodJob] [2888577] [GoodJob::Scheduler(queues=* max_threads=5)-worker-5] Executed GoodJob 3fe6aaca-3718-43f1-bd52-b152c8b78b24
```

I didn't dig into internals yet, but I'm opening the issue as maybe it is already a known issue?

**Comment by bensheldon** [comment]
@morgoth thanks for reporting this 🙏  This is a new issue to me, though reminds me of #117 (which was an issue with Apartment gem).

GoodJob::Job, which is responsible for making connections, inherits from `ActionRecord::Base` which likely doesn't have the the connection info. I don't have any experience with multiple databases though 😞 

Let me bake up a quick PR to see if it can be easily configured to address this.

**Comment by bensheldon** [comment]
@morgoth I just made a PR: https://github.com/bensheldon/good_job/pull/237

Could you see if it unblocks your issue?

One thing that is also concerning is `WARNING:  you don't own a lock of type ExclusiveLock`, which implies that the connection that is unlocking the job didn't actually have a lock on it. I am unsure, with multiple databases, how to ensure that it doesn't switch between multiple database connection pools at all.

**Comment by morgoth** [comment]
@bensheldon Thank you for a quick response.
I tested it and it indeed fixes the problem. It also does not output the warning about exclusive lock.

However this way we're creating another problem with opening separate connection to database.
As explained by Eileen https://github.com/rails/rails/pull/36488/files#diff-a76e37fa9848d683960604d1bd58b37bd4a1993cf5512981a5b2b357eee93698R125-R130 there should be only one parent class for the same database.

So the best fix would be to allow to specify parent class (in my case it would be `ApplicationRecord`), but it might be tricky as when loading the gem code it would already have to be defined.

I looked at the devise gem, where it is possible to specify parent classes for controllers and mailers https://github.com/heartcombo/devise/blob/5d5636f03ac19e8188d99c044d4b5e90124313af/lib/devise.rb#L230-L240
I believe they are able to achieve this by using `autoload` https://github.com/heartcombo/devise/blob/5d5636f03ac19e8188d99c044d4b5e90124313af/lib/devise.rb#L11-L48 , but that's only my assumption - I'm not sure about it.

Maybe that wouldn't be a big problem here with zetiwerk in place. I would need to investigate.

I also tested changing `ActiveRecord::Base` to `ApplicationRecord` in the good job code (just locally) and it works as expected.

**Comment by morgoth** [comment]
Looks like locally something like this works fine https://github.com/bensheldon/good_job/pull/238 
but I didn't try it on production env yet.

**Comment by morgoth** [comment]
I'm testing this in production env (heroku review app). There is no issue with "No connection pool", however I observe many opened connection to pg.

I have started 5 separate instances with good job processes, and configured GOOD_JOB_MAX_THREADS to 1.
```
bundle exec good_job start --poll-interval=1 --queues="default;mailers"
bundle exec good_job start --poll-interval=1 --queues="geocoding"
bundle exec good_job start --poll-interval=1 --queues="almighty"
bundle exec good_job start --poll-interval=1 --queues="tour_planning"
bundle exec good_job start --poll-interval=1 --queues="webhooks"
```
```
 heroku pg:psql -c 'select usename, application_name, query from pg_stat_activity;' --app my-app | grep fjqjvywimdixzv
--> Connecting to postgresql-fluffy-13909
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs"."scheduled_at", "good_jobs"."created_at" FROM "good_jobs" LEFT JOIN pg_locks ON pg_locks.locktype = 'advisory' AND pg_locks.objsubid = 1 AND pg_locks.classid = ('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(32)::int AND pg_locks.objid = (('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(64) << 32)::bit(32)::int WHERE "good_jobs"."queue_name" = $1 AND "pg_locks"."locktype" IS NULL AND "good_jobs"."finished_at" IS NULL AND (scheduled_at < '2021-04-16 12:53:50.578584' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY COALESCE(scheduled_at, created_at) ASC LIMIT $2
 fjqjvywimdixzv | GoodJob::Notifier                          | LISTEN good_job
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'default' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:27.122767' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
 fjqjvywimdixzv | GoodJob::Notifier                          | LISTEN good_job
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs"."scheduled_at", "good_jobs"."created_at" FROM "good_jobs" LEFT JOIN pg_locks ON pg_locks.locktype = 'advisory' AND pg_locks.objsubid = 1 AND pg_locks.classid = ('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(32)::int AND pg_locks.objid = (('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(64) << 32)::bit(32)::int WHERE "good_jobs"."queue_name" = $1 AND "pg_locks"."locktype" IS NULL AND "good_jobs"."finished_at" IS NULL AND (scheduled_at < '2021-04-16 12:53:40.319265' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY COALESCE(scheduled_at, created_at) ASC LIMIT $2
 fjqjvywimdixzv | psql non-interactive                       | select usename, application_name, query from pg_stat_activity;
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'geocoding' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:26.378121' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'mailers' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:27.125149' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'almighty' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:26.415824' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
 fjqjvywimdixzv | GoodJob::Notifier                          | LISTEN good_job
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs"."scheduled_at", "good_jobs"."created_at" FROM "good_jobs" LEFT JOIN pg_locks ON pg_locks.locktype = 'advisory' AND pg_locks.objsubid = 1 AND pg_locks.classid = ('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(32)::int AND pg_locks.objid = (('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(64) << 32)::bit(32)::int WHERE "good_jobs"."queue_name" = $1 AND "pg_locks"."locktype" IS NULL AND "good_jobs"."finished_at" IS NULL AND (scheduled_at < '2021-04-16 12:53:19.360741' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY COALESCE(scheduled_at, created_at) ASC LIMIT $2
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs"."scheduled_at", "good_jobs"."created_at" FROM "good_jobs" LEFT JOIN pg_locks ON pg_locks.locktype = 'advisory' AND pg_locks.objsubid = 1 AND pg_locks.classid = ('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(32)::int AND pg_locks.objid = (('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(64) << 32)::bit(32)::int WHERE "good_jobs"."queue_name" = $1 AND "pg_locks"."locktype" IS NULL AND "good_jobs"."finished_at" IS NULL AND (scheduled_at < '2021-04-16 12:53:32.227795' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY COALESCE(scheduled_at, created_at) ASC LIMIT $2
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'webhooks' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:26.496094' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
 fjqjvywimdixzv | GoodJob::Notifier                          | LISTEN good_job
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs"."scheduled_at", "good_jobs"."created_at" FROM "good_jobs" LEFT JOIN pg_locks ON pg_locks.locktype = 'advisory' AND pg_locks.objsubid = 1 AND pg_locks.classid = ('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(32)::int AND pg_locks.objid = (('x' || substr(md5('good_jobs' || "good_jobs"."id"::text), 1, 16))::bit(64) << 32)::bit(32)::int WHERE "good_jobs"."queue_name" = $1 AND "pg_locks"."locktype" IS NULL AND "good_jobs"."finished_at" IS NULL AND (scheduled_at < '2021-04-16 12:53:18.085628' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY COALESCE(scheduled_at, created_at) ASC LIMIT $2
 fjqjvywimdixzv | GoodJob::Notifier                          | LISTEN good_job
 fjqjvywimdixzv | /app/vendor/bundle/ruby/2.7.0/bin/good_job | SELECT "good_jobs".* FROM "good_jobs" WHERE "good_jobs"."id" IN (WITH "rows" AS  ( SELECT "good_jobs"."id" FROM "good_jobs" WHERE "good_jobs"."queue_name" = 'tour_planning' AND "good_jobs"."finished_at" IS NULL AND ("good_jobs"."scheduled_at" <= '2021-04-16 13:29:26.505474' OR "good_jobs"."scheduled_at" IS NULL) ORDER BY priority DESC NULLS LAST ) SELECT "rows"."id" FROM "rows" WHERE pg_try_advisory_lock(('x' || substr(md5('good_jobs' || "rows"."id"::text), 1, 16))::bit(64)::bigint) LIMIT $1) ORDER BY priority DESC NULLS LAST
```

which gives me 16 connections. Is this normal to have them that many? Why is that?

**Comment by bensheldon** [comment]
@morgoth This is great work 🙌  I reviewed the PR and I'm happy moving forward with that.

> ... which gives me 16 connections. Is this normal to have them that many? Why is that?

I could imagine that's about right if the there are jobs present to execute. 

To break down each process, there should be database connections:

- exactly 1 database connection for LISTEN/NOTIFY. That connection is removed from the Rails connection pool and held open for LISTEN.
- at least 1, and at most the number of execution threads. Job query, lock, execute, cleanup, unlock all happen within the same thread, so here should be _at most_ 1 connection for each execution thread. These are all shared out of the Rails connection pool, so they will only be created if they are needed. For example, in the default size 5 threadpool, GoodJob will wake-up a single thread, the thread will check for work (checking out a connection), and then the thread will go back to sleep (and release the connection back to the pool). If there is work, and the initial thread is engaged and the poll interval comes around again, only then will GoodJob wake up a 2nd thread and create a 2nd database connection.

...but now that I'm reading through those queries you surfaced, I'm realizing that there _is_ an additional query that is happening in the main process: the `GoodJob::Scheduler#warm_cache` is called at startup _in the main process_. That will result in an additional database connection above and beyond the number in the execution thread pool. It should still be part of the DB connection pool, but I'm now thinking that it's not being returned to the pool. That's a bug; it should probably be executed as a task in the execution threadpool so that it doesn't throw off the accounting.

**Comment by jackscotti** [comment]
👋  @bensheldon I get the same failure `[GoodJob] GoodJob error: No connection pool for 'ActiveRecord::Base' found.` when I try to run `good_job` in async_server mode. 

```
Puma caught this error: No connection pool for 'ActiveRecord::Base' found. (ActiveRecord::ConnectionNotEstablished)
/Users/jack/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/activerecord-6.1.3.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
/Users/jack/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/activerecord-6.1.3.1/lib/active_record/connection_handling.rb:327:in `retrieve_connection'
/Users/jack/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/activerecord-6.1.3.1/lib/active_record/connection_handling.rb:283:in `connection'
```

I tried to follow [the instructions ](https://github.com/bensheldon/good_job#execute-jobs-async--in-process) but I have not managed to get it to work.

**Comment by bensheldon** [comment]
@jackscotti thanks for reporting that! Are you using Rails multiple databases too? 

We have what appears to be a working fix in #238. Could you give that branch a try?

```ruby
# Gemfile
gem "good_job", github: " tiramizoo/good_job", branch: "parent-class"

# config/initializers/good_job.rb
GoodJob.active_record_parent_class = "ApplicationRecord" # the ActiveRecord base class that has the database connection info for good_job
```

**Comment by jackscotti** [comment]
@bensheldon thank you for getting back to me so quickly, love this gem!

No multiple database for me. I am working on an almost vanilla Rails app, running Ruby 3.0.0 and Rails 6.1.3.

I have just tried that branch, initially, the server logs look more promising but pages are never loaded and the server does not seem to even receive the requests:

```
➜  git:(development) ✗ rails s
=> Booting Puma
=> Rails 6.1.3.1 application starting in development
=> Run `bin/rails server --help` for more startup options
Puma starting in single mode...
* Puma version: 5.2.2 (ruby 3.0.0-p0) ("Fettisdagsbulle")
*  Min threads: 5
*  Max threads: 5
*  Environment: development
*          PID: 50258
* Listening on http://127.0.0.1:3000
* Listening on http://[::1]:3000
Use Ctrl-C to stop
[GoodJob] [50258] GoodJob started scheduler with queues=* max_threads=5.
```

**Comment by bensheldon** [comment]
@jackscotti nuts. I'll look into it this afternoon. `async_server` was just released this past week and seems like it has some teething issues. 

In the meantime, could you try the long-used workaround to `async`  that was intended to be addressed with `async_server`?

```bash
# remove any assignment of `config.good_job.execution_mode = :async_server, and run the server executable with...
$ GOOD_JOB_EXECUTION_MODE=async_server rails server
```

**Comment by jackscotti** [comment]
Hey @bensheldon just tried as suggested and both good_job 1.9 and the branch shared above do not work on my machine

With goodjob 1.9:
```
Puma caught this error: No connection pool for 'ActiveRecord::Base' found. (ActiveRecord::ConnectionNotEstablished)
```
With `gem "good_job", github: " tiramizoo/good_job", branch: "parent-class"`:
- same issue above with hanging server

**Comment by bensheldon** [comment]
@jackscotti I'm at a loss for why you would be experiencing problems.  If the server is hanging, that could be an indication of an autoloader deadlock. I've seen this happen if something is initializing ActiveRecord or ActiveJob prematurely in the Rails initialization process.

Here's my calendar if you wanted to hop on Zoom call and we could try to debug it: https://calendly.com/ben-cfa/chat

**Comment by bensheldon** [comment]
If you wanted to try to diagnose the deadlock yourself:

```ruby
# Add to your gemfile
gem 'rbtrace'

# Create an initializer: config/initializers/rbtrace.rb
if ENV['RBTRACE']
  require 'rbtrace'
  sleep 1
  $stdout.puts "Enabled rbtrace on PID #{Process.pid}. Example commands to run from a different terminal:"
  $stdout.puts "  - Show all method calls: $ bundle exec rbtrace --pid #{Process.pid} --firehose"
  $stdout.puts "  - Debug Rails deadlock: $ bundle exec rbtrace --pid #{Process.pid} --eval \"puts output = ActionDispatch::DebugLocks.new(nil).send(:render_details, nil); output\""
  $stdout.puts 'Press Enter to continue loading...'
  $stdin.gets
end

# Start your puma process in one terminal
# ...and in a second terminal, run those commands.
```

**Comment by jackscotti** [comment]
hey @bensheldon , sorry for disappearing. Been quite busy in the past days, you seem to be US based so I can't easily book a call with you. I am going to give another try at debugging it myself this weekend.

**Comment by bensheldon** [comment]
@jackscotti did you make any progress over the weekend? I'm usually up and available by 7am Pacific (14 UTC) if you can suggest a day.

I don't think Webpacker would be a problem. I'm using GoodJob in my own website that uses Webpacker.

**Comment by bensheldon** [comment]
@jackscotti I was able to reproduce the deadlock 🎉  I am working on a fix: #242

**Comment by jackscotti** [comment]
Hi Ben, I am sorry for not having helped enough but I am super happy you have have managed to reproduce the problem. I will put my best effort to try to give back to this project!
